syntax = "proto3";

package auth;

message LoginRequest {
  string username = 1;
}

message LoginResponse {
  enum LoginStep {
    NEED_PASSWORD = 0;
    SENT_LOGIN_EMAIL = 1;
    LOGIN_NO_SUCH_USER = 3;
  }

  LoginStep next_step = 1;
}

message SignupRequest {
  string email = 1;
}

message SignupResponse {
  enum SignupStep {
    SENT_SIGNUP_EMAIL = 0;
    EMAIL_EXISTS = 1;
  }

  SignupStep next_step = 1;
}

message AuthRequest {
  string username = 1;
  string password = 2;
}

message AuthResponse {
  string token = 1;
}

message DeauthRequest {
  string token = 1;
}

message DeauthResponse {}

service Auth {
  rpc Login(LoginRequest) returns (LoginResponse);
  rpc Signup(SignupRequest) returns (SignupResponse);

  rpc Authenticate(AuthRequest) returns (AuthResponse);
  rpc Deauthenticate(DeauthRequest) returns (DeauthResponse);
}
